### PR TITLE
Exit and cycle

### PR DESCRIPTION
**Pull request by richardblam**
I believe this is an improvement on the exit and cycle statements for loops and blocks. It handles exiting and cycling from a loop or block to the parent named statements in llvm. There are a couple of caveats and things I couldn't quit figure out that I could use feedback on.
- I thought the block name could be optional but I had a problem with tests/block1.f90 without creating the block name "block" as it was.
- I don't really understand the asr verification require(symtab_in_scope(current_symtab, x.m_m)

Overall, it works quite well with the new cycle_and_exit tests. I thought I should probably get feedback without continuing further since it touches a lot of files and I was slightly confused about the existing tests/block1.f90 test that has a symbol table that I don't fully grasp now. The other changes are mostly for having a loop have a name so it can be found in the codegen for branching.

### CODE REVIEW

**Comment by richardblam** [comment]
I have fixed up this branch so it has more minimal changes vs. the original. The issues I had are addressed as follows.
- The block name is not optional as it was so the symbol table behavior is unchanged
- Verify asr such as require(symtab_in_scope(current_symtab, x.m_m) is unchanged

One explanation to understand this pull request is that "exit stmt_name" is the same for both loops and blocks and simply branches to either a block or loop end which are both put on one stack. "cycle stmt_name" is only for loops.

**Comment by richardblam** [comment]
The goal of this pull request is to close issue #1414.

**Comment by richardblam** [comment]
I would be happy to test the particular block exit to named statement in the modern minpack if someone pointed me to where that particular problem was or if someone else has that available quickly they could try it with this branch.

**Comment by certik** [comment]
Thanks @richardblam !!

@Pranavchiku can you test this out and review it?

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 1]
```diff
@@ -63,17 +59,25 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+        ASR::asr_t* block;
+        std::string name;
+        if (x.m_stmt_name) {
+            name = std::string(x.m_stmt_name);
+            block = ASR::make_Block_t(al, x.base.base.loc,
+                                      current_scope, x.m_stmt_name,
+                                      body.p, body.size());
+        } else {
+            // TODO: Understand tests/block1.f90 to know if this is needed, otherwise
+            // it might be possible to allow x.m_stmt_name to be nullptr
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_body_visitor.cpp:71]: Yes, it is needed, previously we treated all block names to be "block" by neglecting x.m_stmt_name.

**Review by czgdp1807 — APPROVED** [review]
I checked the integration tests and they appear to be strong enough. So if they are passing with this change on the CI, then I think its good.

**Comment by Pranavchiku** [comment]
I am able to see divergence in modern_minpack/examples verifying if it is because of this PR against the latest main.

**Review by Pranavchiku — APPROVED** [review]
Looks good to me! There is some divergence with the examples that aligned with gfortran in latest main will fix that later on.

File: tests/tests.toml [hunk 2]
```diff
@@ -77,6 +77,30 @@ ast = true
...
+filename = "cycle_and_exit3.f90"
+ast = true
+asr = true
+llvm = true
+
+[[test]]
+filename = "cycle_and_exit4.f90"
+ast = true
+asr = true
+llvm = true
```
Comment (certik) [inline tests/tests.toml:102]: I think we don't need all these in here. The AST seems to be working, so we don't need to test it.

ASR we can test, if needed.

Do these tests run correctly? If so, why not adding them to integration_tests and just run them?

Comment (richardblam) [reply tests/tests.toml:102]: I moved the tests to the integration_tests folder. I still have the tests listed in tests.toml. Do they then get removed from tests.toml?

**Comment by Pranavchiku** [comment]
Yeah, you may remove those.

File: tests/tests.toml (hunk 2 above)
Comment (richardblam) [reply tests/tests.toml]: Ok, I removed the tests from tests.toml now.

**Comment by richardblam** [comment]
What ends up staying as far as the reference output?

File: src/lfortran/semantics/ast_body_visitor.cpp (hunk 1 above)
Comment (richardblam) [reply src/lfortran/semantics/ast_body_visitor.cpp:71]: Ok, it now uses x.m_stmt_name or the name "block" if it's not present.

**Comment by richardblam** [comment]
I merged the main branch to get this ready.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 3]
```diff
@@ -5341,20 +5366,60 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
 
-    void visit_Exit(const ASR::Exit_t & /* x */) {
-        builder->CreateBr(do_loop_end.back());
+    void visit_Exit(const ASR::Exit_t &x) {
+        if (x.m_stmt_name) {
+            std::string stmt_name = std::string(x.m_stmt_name) + ".end";
+            int nested_block_depth = loop_or_block_end_names.size();
+            int i = nested_block_depth - 1;
+            for (; i >= 0; i--) {
+                if (case_insensitive_string_compare(loop_or_block_end_names[i], stmt_name)) {
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:5387]: We shouldn't need to do this, rather we should convert to lowercase when saving into ASR.

Also, we should actually already figure out ahead of time which loop is tied to this Exit. We should not be figuring this out in all the backends.

**Review by certik — APPROVED** [review]
I think this is good enough to be merged. Thanks @richardblam for fixing this.

I left some comments that can be fixed in subsequent PRs.

I think it's good to store the labels in ASR, so that when we print ASR back to Fortran (on my TODO list!), it prints the original labels. For backend purposes, we should not be handling strings at all, we should somehow denote in ASR which exact loop corresponds to the given Exit, perhaps via an integer offset, and just use it. The idea is for the backend to do minimal amount of this "semantic" work. This can be improved later.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 4]
```diff
@@ -63,17 +59,25 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
-        std::string name = parent_scope->get_unique_name("block");
-        ASR::asr_t* block = ASR::make_Block_t(al, x.base.base.loc,
-                                              current_scope, s2c(al, name),
-                                              body.p, body.size());
+        ASR::asr_t* block;
+        std::string name;
+        if (x.m_stmt_name) {
+            name = std::string(x.m_stmt_name);
+            block = ASR::make_Block_t(al, x.base.base.loc,
+                                      current_scope, x.m_stmt_name,
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:67]:
```suggestion
                                      current_scope, to_lower(x.m_stmt_name),
```

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 5]
```diff
@@ -2322,7 +2326,7 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         Vec<ASR::stmt_t*> body;
         body.reserve(al, x.n_body);
         transform_stmts(body, x.n_body, x.m_body);
-        tmp = ASR::make_WhileLoop_t(al, x.base.base.loc, test, body.p,
+        tmp = ASR::make_WhileLoop_t(al, x.base.base.loc, x.m_stmt_name, test, body.p,
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:2329]:
```suggestion
        tmp = ASR::make_WhileLoop_t(al, x.base.base.loc, to_lower(x.m_stmt_name), test, body.p,
```

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 6]
```diff
@@ -2365,14 +2369,14 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         head.m_increment = increment;
         if (head.m_v != nullptr) {
             head.loc = head.m_v->base.loc;
-            tmp = ASR::make_DoLoop_t(al, x.base.base.loc, head, body.p, body.size());
+            tmp = ASR::make_DoLoop_t(al, x.base.base.loc, x.m_stmt_name, head, body.p, body.size());
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:2372]: to_lower

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 7]
```diff
@@ -2365,14 +2369,14 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
             do_loop_variables.pop_back();
         } else {
             ASR::ttype_t* cond_type
                 = ASRUtils::TYPE(ASR::make_Logical_t(al, x.base.base.loc, 4, nullptr, 0));
             ASR::expr_t* cond = ASRUtils::EXPR(
                 ASR::make_LogicalConstant_t(al, x.base.base.loc, true, cond_type));
-            tmp = ASR::make_WhileLoop_t(al, x.base.base.loc, cond, body.p, body.size());
+            tmp = ASR::make_WhileLoop_t(al, x.base.base.loc, x.m_stmt_name, cond, body.p, body.size());
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:2379]: to_lower

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 8]
```diff
@@ -2466,20 +2470,11 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
-        // TODO: add a check here that we are inside a While loop
-        tmp = ASR::make_Exit_t(al, x.base.base.loc);
-        if (x.m_stmt_name) {
-            std::string exit_label = x.m_stmt_name;
-            // Check if it is an exit from block, if so, create an ExitBlock
-            if (block_names.size() > 0 && std::find(block_names.begin(), block_names.end(), exit_label) != block_names.end()) {
-                tmp = ASR::make_ExitBlock_t(al, x.base.base.loc);
-            }
-        }
+        tmp = ASR::make_Exit_t(al, x.base.base.loc, x.m_stmt_name);
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:2473]: to_lower

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 9]
```diff
@@ -2466,20 +2470,11 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
     }
 
     void visit_Cycle(const AST::Cycle_t &x) {
-        // TODO: add a check here that we are inside a While loop
-        tmp = ASR::make_Cycle_t(al, x.base.base.loc);
+        tmp = ASR::make_Cycle_t(al, x.base.base.loc, x.m_stmt_name);
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:2477]: to_lower

**Comment by certik** [comment]
@richardblam don't push into this branch, I am waiting for the CI to finish, so that we can merge it.

**Comment by richardblam** [comment]
Thanks for fixing the integration tests. Showing the design philosophy for case insensitivity is very helpful.

**Comment by certik** [comment]
> Showing the design philosophy for case insensitivity is very helpful.

One issue that I can see with our future ASR -> Fortran backend is that all upper case variable names will be converted to lowercase. An alternative design would be to store in ASR the exact spelling from the declaration, and use case insensitive lookup for variable names (we use a hashtable, so probably the keys should be lowercase in either case). As a user I want the compiler to warn me for inconsistent uppercase/lowercase spelling. But overall it's tricky, we would need to resolve all corner cases (say using legacy uppercase Fortran code), etc. I think we can tackle this issue once we can compile most Fortran codes and have more real world tests at our CI.
